src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/BaseHeartbeatProperties.cs

```diff
@@ -0,0 +1,181 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using System.Text;
+    using System.Threading.Tasks;
+#if NETSTANDARD1_3
+    using System.Runtime.InteropServices;
+#endif
+
+    internal class BaseHeartbeatProperties : IHeartbeatDefaultPayloadProvider
+    {
+        internal readonly List<string> DefaultFields = new List<string>()
+        {
+            // "runtimeFramework",
+            "baseSdkTargetFramework",
+            "osType",
+            "processSessionId"
+        };
+
+        /// <summary>
+        /// A unique identifier that would help to indicate to the analytics when the current process session has
+        /// restarted. 
+        /// 
+        /// <remarks>If a process is unstable and is being restared frequently, tracking this property
+        /// in the heartbeat would help to identify this unstability.
+        /// </remarks>
+        /// </summary>
+        private static Guid? uniqueProcessSessionId = null;
+
+        public string Name => "Base";
+
+        public bool IsKeyword(string keyword)
+        {
+            return this.DefaultFields.Contains(keyword, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public Task<bool> SetDefaultPayload(IEnumerable<string> disabledFields, IHeartbeatProvider provider)
+        {
+            bool hasSetValues = false;
+            var enabledProperties = this.DefaultFields.Except(disabledFields);
+
+            foreach (string fieldName in enabledProperties)
+            {
+                // we don't need to report out any failure here, so keep this look within the Sdk Internal Operations as well
+                try
+                {
+                    switch (fieldName)
+                    {
+                        // case "runtimeFramework":
+                        //    provider.AddHeartbeatProperty(fieldName, true, this.GetRuntimeFrameworkVer(), true);
+                        //    hasSetValues = true;
+                        //    break;
+                        case "baseSdkTargetFramework":
+                            provider.AddHeartbeatProperty(fieldName, true, this.GetBaseSdkTargetFramework(), true);
```

Comment:
I think the AddHeartbeatProperty with 4 args is the only one used from IHeartbeatPropertyManager. It still contains AddHeartbeatProperty with 3 args.

Reply:
Internally, yes, this is the only one used (as in the base SDK we allow the overriding of default properties). Externally, we only allow adding properties that do not collide with the ‘default’ property names. See my PR in -dotnet-server for an example of the use of the `AddHeartbeatProperty` with 3 args.